private/pkg/storage/storagegit/bucket.go

```diff
@@ -0,0 +1,202 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storagegit
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+
+	"github.com/bufbuild/buf/private/pkg/git"
+	"github.com/bufbuild/buf/private/pkg/normalpath"
+	"github.com/bufbuild/buf/private/pkg/storage"
+	"github.com/bufbuild/buf/private/pkg/storage/storageutil"
+)
+
+type bucket struct {
+	objectReader git.ObjectReader
+	symlinks     bool
+	root         git.Tree
+}
+
+func newBucket(
+	objectReader git.ObjectReader,
+	symlinksIfSupported bool,
+	root git.Tree,
+) (storage.ReadBucket, error) {
+	return &bucket{
+		objectReader: objectReader,
+		symlinks:     symlinksIfSupported,
+		root:         root,
+	}, nil
+}
+
+func (b *bucket) Get(ctx context.Context, path string) (storage.ReadObjectCloser, error) {
+	node, err := b.root.Traverse(b.objectReader, normalpath.Components(path)...)
+	if err != nil {
+		if errors.Is(err, git.ErrSubTreeNotFound) {
+			return nil, storage.NewErrNotExist(path)
+		}
+		return nil, err
+	}
+	switch node.Mode() {
+	case git.ModeFile, git.ModeExe:
+		data, err := b.objectReader.Blob(node.Hash())
+		if err != nil {
+			return nil, err
+		}
+		return &namedReader{
+			info:   b.newObjectInfo(path),
+			reader: bytes.NewReader(data),
+		}, nil
+	case git.ModeSymlink:
+		if !b.symlinks {
+			return nil, storage.NewErrNotExist(path)
+		}
+		// Symlinks are stored as blobs that reference the target path as a relative
+		// path. We can follow this symlink trivially.
+		data, err := b.objectReader.Blob(node.Hash())
+		if err != nil {
+			return nil, err
+		}
+		path = normalpath.Join(
+			normalpath.Base(path),
+			normalpath.Normalize(string(data)),
+		)
+		return b.Get(ctx, path)
+	default:
+		return nil, storage.NewErrNotExist(path)
+	}
+}
+
+func (b *bucket) Stat(ctx context.Context, path string) (storage.ObjectInfo, error) {
+	node, err := b.root.Traverse(b.objectReader, normalpath.Components(path)...)
+	if err != nil {
+		if errors.Is(err, git.ErrSubTreeNotFound) {
+			return nil, storage.NewErrNotExist(path)
+		}
+		return nil, err
+	}
+	switch node.Mode() {
+	case git.ModeFile, git.ModeExe:
+		return b.newObjectInfo(path), nil
+	case git.ModeSymlink:
+		if !b.symlinks {
+			return nil, storage.NewErrNotExist(path)
+		}
+		return b.newObjectInfo(path), nil
+	default:
+		// TODO: should this be an error? What kind of error can we throw here?
+		return nil, storage.NewErrNotExist(path)
+	}
+}
+
+func (b *bucket) Walk(ctx context.Context, prefix string, f func(storage.ObjectInfo) error) error {
+	walkChecker := storageutil.NewWalkChecker()
+	return b.walk(b.root, b.objectReader, prefix, func(path string, te git.Node) error {
```

Comment:
```suggestion
	return b.walk(b.root, b.objectReader, prefix, func(path string, _ git.Node) error {
```

or even further, do you actually need to pass the node to the walk func?
```suggestion
	return b.walk(b.root, b.objectReader, prefix, func(path string) error {
```

Reply:
You're right, I don't. Let me clean up.